src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts

```diff
@@ -88,10 +121,64 @@ export const changeTrackPanelModalViewForGenome: ActionCreator<
   );
 };
 
+export const updatePreviouslyViewedObjectsAndSave: ActionCreator<
+  ThunkAction<void, any, null, Action<string>>
+> = () => (dispatch, getState: () => RootState) => {
+  const state = getState();
+  const activeGenomeId = getBrowserActiveGenomeId(state);
+  const activeEnsObject = getBrowserActiveEnsObject(state);
+  if (!activeGenomeId || !activeEnsObject) {
+    return;
+  }
+  const trackStates = getBrowserTrackStates(state)[activeGenomeId];
+
+  const activeGenomePreviouslyViewedObjects = [
+    ...getActiveGenomePreviouslyViewedObjects(state)
+  ];
+
+  const existingIndex = activeGenomePreviouslyViewedObjects.findIndex(
+    (previouslyViewedObject) =>
+      previouslyViewedObject.object_id === activeEnsObject.object_id
+  );
+  if (existingIndex === -1) {
+    // IF it is not present, add it to the end
+    activeGenomePreviouslyViewedObjects.push(
+      buildBookmark(activeEnsObject, trackStates)
+    );
+  } else if (existingIndex !== -1) {
+    // If it is already present, bump it to the end
+    activeGenomePreviouslyViewedObjects.push({
+      ...activeGenomePreviouslyViewedObjects[existingIndex],
+      trackStates
+    });
+    activeGenomePreviouslyViewedObjects.splice(existingIndex, 1);
+  }
```

Comment:
1) `else if` -> `else`
2) in order to avoid actually creating a new previously viewed object, would this approach work better?

```
const [previouslyViewedObject] = activeGenomePreviouslyViewedObjects.splice(existingIndex, 1);
activeGenomePreviouslyViewedObjects.push(previouslyViewedObject);
```
?

As a side note, the `activeGenome` part of `activeGenomePreviouslyViewedObjects` name looks completely unnecessary. Could we drop it?

Reply:
Done these changes.